File: apple-ds-core/03-哈希表/hash_table.py
"""
哈希表实现（链地址法）
"""

class HashTable:
    """哈希表实现"""
    
    def __init__(self, capacity=16):
        """初始化
        
        Args:
            capacity: 初始容量
        """
        self.capacity = capacity
        self.size = 0
        self.table = [[] for _ in range(capacity)]
        self.load_factor_threshold = 0.75
    
    def _hash(self, key):
        """哈希函数
        
        Args:
            key: 键（支持字符串和整数）
            
        Returns:
            哈希值
        """
        if isinstance(key, str):
            hash_val = 0
            for char in key:
                hash_val = (hash_val * 31 + ord(char)) % self.capacity
            return hash_val
        else:
            return hash(key) % self.capacity
    
    def _resize(self):
        """扩容并重新哈希"""
        old_table = self.table
        self.capacity *= 2
        self.table = [[] for _ in range(self.capacity)]
        self.size = 0
        
        for bucket in old_table:
            for key, value in bucket:
                self.insert(key, value)
    
    def insert(self, key, value):
        """插入键值对
        
        Args:
            key: 键
            value: 值
        """
        # 检查是否需要扩容
        if self.size / self.capacity > self.load_factor_threshold:
            self._resize()
        
        index = self._hash(key)
        bucket = self.table[index]
        
        # 更新已存在的key
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket[i] = (key, value)
                return
        
        # 插入新key
        bucket.append((key, value))
        self.size += 1
    
    def get(self, key):
        """获取值
        
        Args:
            key: 键
            
        Returns:
            值
            
        Raises:
            KeyError: 键不存在
        """
        index = self._hash(key)
        bucket = self.table[index]
        
        for k, v in bucket:
            if k == key:
                return v
        
        raise KeyError(f'Key {key} not found')
    
    def remove(self, key):
        """删除键值对
        
        Args:
            key: 键
            
        Raises:
            KeyError: 键不存在
        """
        index = self._hash(key)
        bucket = self.table[index]
        
        for i, (k, v) in enumerate(bucket):
            if k == key:
                del bucket[i]
                self.size -= 1
                return
        
        raise KeyError(f'Key {key} not found')
    
    def contains(self, key):
        """检查键是否存在
        
        Args:
            key: 键
            
        Returns:
            bool
        """
        try:
            self.get(key)
            return True
        except KeyError:
            return False
    
    def keys(self):
        """返回所有键"""
        result = []
        for bucket in self.table:
            for k, v in bucket:
                result.append(k)
        return result
    
    def values(self):
        """返回所有值"""
        result = []
        for bucket in self.table:
            for k, v in bucket:
                result.append(v)
        return result
    
    def items(self):
        """返回所有键值对"""
        result = []
        for bucket in self.table:
            for item in bucket:
                result.append(item)
        return result
    
    def __len__(self):
        return self.size
    
    def __str__(self):
        items = [f"{k}: {v}" for k, v in self.items()]
        return "{" + ", ".join(items) + "}"


def demo():
    """演示哈希表的使用"""
    print("=== 哈希表演示 ===\n")
    
    ht = HashTable(4)
    
    # 插入
    print("插入键值对:")
    ht.insert("apple", 5)
    print(f"  apple: 5")
    ht.insert("banana", 3)
    print(f"  banana: 3")
    ht.insert("orange", 7)
    print(f"  orange: 7")
    
    print(f"\n哈希表: {ht}")
    print(f"大小: {len(ht)}, 容量: {ht.capacity}\n")
    
    # 查找
    print(f"查找 'apple': {ht.get('apple')}")
    print(f"包含 'grape': {ht.contains('grape')}\n")
    
    # 更新
    print("更新 'apple' 为 10")
    ht.insert("apple", 10)
    print(f"哈希表: {ht}\n")
    
    # 继续插入触发扩容
    print("继续插入触发扩容:")
    ht.insert("grape", 8)
    ht.insert("melon", 6)
    print(f"哈希表: {ht}")
    print(f"大小: {len(ht)}, 容量: {ht.capacity}\n")
    
    # 删除
    print("删除 'banana'")
    ht.remove("banana")
    print(f"哈希表: {ht}\n")
    
    # 遍历
    print("所有键:", ht.keys())
    print("所有值:", ht.values())


if __name__ == '__main__':
    demo()

File: apple-ds-core/03-哈希表/hash_table.cpp
/**
 * 哈希表实现（链地址法）
 */

#include <iostream>
#include <vector>
#include <list>
#include <string>
#include <stdexcept>

template<typename K, typename V>
class HashTable {
private:
    struct Entry {
        K key;
        V value;
        Entry(K k, V v) : key(k), value(v) {}
    };
    
    std::vector<std::list<Entry>> table;
    int size;
    int capacity;
    double load_factor_threshold;
    
    int hash(const K& key) const {
        return std::hash<K>{}(key) % capacity;
    }
    
    void resize() {
        int old_capacity = capacity;
        capacity *= 2;
        
        std::vector<std::list<Entry>> old_table = table;
        table = std::vector<std::list<Entry>>(capacity);
        size = 0;
        
        for (const auto& bucket : old_table) {
            for (const auto& entry : bucket) {
                insert(entry.key, entry.value);
            }
        }
    }

public:
    HashTable(int cap = 16) 
        : capacity(cap), size(0), load_factor_threshold(0.75) {
        table.resize(capacity);
    }
    
    void insert(const K& key, const V& value) {
        if ((double)size / capacity > load_factor_threshold) {
            resize();
        }
        
        int index = hash(key);
        auto& bucket = table[index];
        
        // 更新已存在的key
        for (auto& entry : bucket) {
            if (entry.key == key) {
                entry.value = value;
                return;
            }
        }
        
        // 插入新key
        bucket.push_back(Entry(key, value));
        size++;
    }
    
    V get(const K& key) const {
        int index = hash(key);
        const auto& bucket = table[index];
        
        for (const auto& entry : bucket) {
            if (entry.key == key) {
                return entry.value;
            }
        }
        
        throw std::runtime_error("Key not found");
    }
    
    void remove(const K& key) {
        int index = hash(key);
        auto& bucket = table[index];
        
        for (auto it = bucket.begin(); it != bucket.end(); ++it) {
            if (it->key == key) {
                bucket.erase(it);
                size--;
                return;
            }
        }
        
        throw std::runtime_error("Key not found");
    }
    
    bool contains(const K& key) const {
        try {
            get(key);
            return true;
        } catch (...) {
            return false;
        }
    }
    
    int getSize() const {
        return size;
    }
    
    void print() const {
        std::cout << "{";
        bool first = true;
        for (const auto& bucket : table) {
            for (const auto& entry : bucket) {
                if (!first) std::cout << ", ";
                std::cout << entry.key << ": " << entry.value;
                first = false;
            }
        }
        std::cout << "}" << std::endl;
    }
};


int main() {
    std::cout << "=== 哈希表演示 ===" << std::endl << std::endl;
    
    HashTable<std::string, int> ht(4);
    
    std::cout << "插入键值对:" << std::endl;
    ht.insert("apple", 5);
    std::cout << "  apple: 5" << std::endl;
    ht.insert("banana", 3);
    std::cout << "  banana: 3" << std::endl;
    ht.insert("orange", 7);
    std::cout << "  orange: 7" << std::endl;
    
    std::cout << "\n哈希表: ";
    ht.print();
    std::cout << "大小: " << ht.getSize() << std::endl << std::endl;
    
    std::cout << "查找 'apple': " << ht.get("apple") << std::endl;
    std::cout << "包含 'grape': " << (ht.contains("grape") ? "是" : "否") << std::endl << std::endl;
    
    std::cout << "更新 'apple' 为 10" << std::endl;
    ht.insert("apple", 10);
    std::cout << "哈希表: ";
    ht.print();
    std::cout << std::endl;
    
    std::cout << "删除 'banana'" << std::endl;
    ht.remove("banana");
    std::cout << "哈希表: ";
    ht.print();
    
    return 0;
}

